DiscordBot.DataAccess/EventsSheetsService.cs

```diff
@@ -83,6 +118,39 @@ public async Task WriteRow()
 
         public async Task AddEventAsync(string name, string description, DateTime time)
         {
+            // Allocate new key
+            largestKey++;
+
+            var newRow = new ValueRange
+            {
+                Values = new IList<object>[]
+                {
+                    new object[]
+                    {
+                        largestKey,
+                        name,
+                        description,
+                        time.ToString("s"),
+                        "Europe/London"
+                    }
+                }
+            };
+
+            var request = sheetsService.Spreadsheets.Values.Append(
+                newRow,
+                spreadsheetId,
+                "A:E"
+            );
+            request.ValueInputOption = AppendRequest.ValueInputOptionEnum.RAW;
+
+            var response = await request.ExecuteAsync();
+
+            // Record which row the new event was inserted on
+            keyToRowNumberMap.Add(
```

Comment:
Is there any disadvantages with having empty rows in the sheet? Because having to keep track of this mapping doesn't seem ideal. Or even if we don't leave empty rows is it not possible to edit/delete rows based on key rather than row number?

And either way this doesn't seem necessary for the functionality right now which just allows inserting? Try to not future proof too much, instead focus on getting a simple neat implementation of the current functionality. That way you're more likely to end up with a simple and neat implementation in the end 🙂

Reply:
I have also been implementing the edit event and delete event methods through the course of the day, but I only thought my add event method was ready to go up for review tonight. Editing events was what drove me to add this functionality, but that method is a bit messy at the moment and I wasn't planning to put it up for review until tomorrow morning.

If we don't keep track of row numbers (either implicitly when row numbers correspond to the key, or explicitly using a dictionary like the current implementation) then to edit a row we must first search through the sheet to find the row the event is being stored in. This didn't seem ideal to me because we could be maintaining this state every time we perform an operation, and we are limited to 10 API calls a second to Google Sheets and a search and then update would require 2 API calls instead of 1. I think this might be worse than keeping track of row numbers in some way, but I could be wrong 🙂

Leaving empty rows would get around this, because we know which row an event is on just by its key number. I started with this, and then to ensure that we could only edit events that hadn't been deleted I started keeping a set much like the keys in that dictionary. At that point I then converted it over to track event key-row mappings, but I could go back to keeping a set of valid event keys if you think it's better? 🙂 Alternatively we could just allow a deleted event to be edited (inserting data into part of a row) but I thought that might not be for the best solution.

Reply:
Search the update seems best to me in terms of code quality and avoiding bugs. Because if we're storing the state in the google sheet then we don't really want to also store some of it ourselves, that adds complexity and adds the possibility that they would get out of sync and cause bugs. 

Do you think we're likely to go over the 10 api calls a second limit? 🙂

Reply:
Ok, I'll switch over to using that then! I am also keeping track of the largest key in use (to know what key to give out to the next event being added), should I switch that over to search and update as well to keep things consistent? 🙂

I think the number of API calls will eventually become a problem, I've had another look and I think the limit will be 100 calls per 100 seconds (see [here](https://developers.google.com/sheets/api/limits)), because we only have one user interacting with the sheet (rather than the 500 calls/100s for the overall project). Depending how we implement multi-role signups, this could end up being 2 calls (search/update) per signup with users signing up for multiple things in quick succession, so an initial rush after publishing an event could overwhelm the API call quota from time to time.

However, having thought about it a little more though, I don't think we're going to be able to avoid the problem just by trimming the odd API call here and there (like I was trying to do here). I think we might need some kind of wait/retry system if we overload our quota for a given 100s to wait until we're allowed to use it again, or look at buying a bigger API quota from Google. Another approach would be to coalesce requests together (you are allowed to submit large numbers of requests in a single batch and it only counts as one call) but that could be a bit complicated to implement.

Reply:
The largest key seems a lot simpler to keep track of, it's just the last entered key right? So I don't really mind, I think either is fine 🙂 

Sounds like we've got a few options then. Submitting requests in batch sounds good, if we can get that working. But I think we should just go with this simple solution for now and just bear in mind that this is something we'll need to deal with at some point 🙂

Reply:
Done this in bf55cbc 🙂